block_synchronizer: add a new final state and fix signature sufficiency determination for blocks with pre-1.5 protocol versions

alsrdn opened the pull request:
Fixes: https://github.com/casper-network/casper-node/issues/3824

`HaveStrictFinality` was considered the terminal state when synchronizing a block even though this was not entirely true. Because on the forward sync flow the synchronizer needed to wait until the block was executed to establish that the sync was finished, there was some unnecessary complexity added to keep `HaveStrictFinality` as a terminal state which involved more complex checks in other intermediary states like `HaveBlock`, `HaveAllDeploys` etc.
The same can be said about the historical sync flow where the synchronizer needed to wait for the block to be marked complete.
This PR introduces a new terminal block acquisition state called `Complete` which fully denotes that the block synchronizer is finished with all tasks involving the block, on both forward and historical sync.
This simplifies some state transitions since enqueuing the block for execution or requesting the block to be marked complete is only done from the `HaveStrictFinality` state.
For the forward sync case, there is another intermediary state called `HaveFinalizedBlock`; the block builder transitions to this state after the block has been successfully converted into a block that is ready for execution.

Furthermore, we now allow sync to progress for historical blocks that were created with a protocol version that is older than the one where the requirement for nodes to hold signatures for strict finality was introduced.
In mainnet this is protocol version 1.5.
For other networks a configurable chainspec parameter was added to control the version when this requirement was introduced.
Because pre-1.5 there was no requirement for nodes to retain finality signatures, there may exist networks were all finality signatures for a block are not available in order to reach weak or strict finality.
Because we have 2 checkpoints for determining if we can progress with the sync (first achieving weak finality, and then strict finality) we employ the `LegacyRequiredFinality` the requirement for the block that is within the criteria described above. Previous code did not take into account the check for weak finality.

alsrdn committed:
block_synchronizer: add new terminal state

`HaveStrictFinality` was considered the terminal state when
synchronizing a block even though this was not entirely true. Because on
the forward sync flow the synchronizer needed to wait until the block
was executed to establish that the sync was finished, there was some
unnecessary complexity added to keep `HaveStrictFinality` as a terminal
state which involved more complex checks in other intermediary states
like `HaveBlock`, `HaveAllDeploys` etc. The same can be said about the
historical sync flow where the synchronizer needed to wait for the block
to be marked complete.
This commit introduces a new terminal block acquisition state called
`Complete` which fully denotes that the block synchronizer is finished
with all tasks involving the block, on both forward and historical sync.
This simplifies some state transitions since enqueuing the block for
execution or requesting the block to be marked complete is only done
from the `HaveStrictFinality` state. For the forward sync case, there is
another intermediary state called `HaveFinalizedBlock`; the block
builder transitions to this state after the block has been successfully
converted into a block that is ready for execution.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
synchronizer: allow progress for pre-1.5 blocks without weak finality

Allow sync to progress for historical blocks that were created with a
protocol version that is older than the one where the requirement for
nodes to hold signatures for strict finality was introduced. In mainnet
this is protocol version 1.5. For other networks a configurable
chainspec parameter was added to control the version when this
requirement was introduced.
Because pre-1.5 there was no requirement for nodes to retain finality
signatures, there may exist networks were all finality signatures for a
block are not available in order to reach weak or strict finality.
Because we have 2 checkpoints for determining if we can progress with
the sync (first achieving weak finality, and then strict finality) we
employ the `LegacyRequiredFinality` the requirement for the block that
is within the criteria described above. Previous code did not take into
account the check for weak finality.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
chainspec: fix ser/deserialization of newly added parameter

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

rafal-ch commented on node/src/components/block_synchronizer/block_acquisition.rs:
```diff
@@ -509,6 +574,47 @@ impl BlockAcquisitionState {
     }
 
     /// Register a finality signature as pending for this block.
+    pub(super) fn switch_to_have_strict_finality(
```
Looks like this comment and the one for the function below need to switch places.

alsrdn replied:
fixed as part of: https://github.com/casper-network/casper-node/pull/3832/commits/fc3e6afc1098cca19300d76e87f4595853042470

rafal-ch commented on node/src/components/block_synchronizer/block_acquisition.rs:
```diff
@@ -438,6 +495,10 @@ impl BlockAcquisitionState {
                         header.height()
                     );
+                    if is_historical && header.protocol_version() < strict_finality_protocol_version
+                    {
```
This `{` in a new line reminds me of the code-formatting wars :-)

alsrdn replied:
fixed as part of: https://github.com/casper-network/casper-node/pull/3832/commits/fc3e6afc1098cca19300d76e87f4595853042470

rafal-ch commented on node/src/components/block_synchronizer/signature_acquisition.rs:
```diff
@@ -93,21 +100,45 @@ impl SignatureAcquisition {
+    // sufficiency:
+    //      * false means that the caller considers `Weak` finality as sufficient
+    //      * true means that the caller considers `Strict` finality as sufficient
+    pub(super) fn has_sufficient_finality(
```
Can you cover this function with UT?

rafal-ch replied:
This is covered in a separate PR: https://github.com/casper-network/casper-node/pull/3842

alsrdn committed:
block_synchronizer: prevent enqueuing block for exec multiple times

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
Merge branch 'dev' into synchronizer-add-new-final-state

alsrdn committed:
block_synchronizer/tests: adjust tests for the new end state

Since new states were introduced for the block acquisition, update the
tests to take them into account.
Alse re-enable some ignored tests since the bugs were fixed.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
block_synchronizer: fix lint errors

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

rafal-ch commented:
bors r+

Commit:
Merge remote-tracking branch 'upstream/dev' into synchronizer-add-new-final-state

georgepisaltu commented on node/src/components/block_synchronizer.rs:
```diff
@@ -1252,15 +1252,29 @@ impl BlockSynchronizer {
+            BlockSynchronizerProgress::Syncing(
+                builder.block_hash(),
+                builder.block_height(),
+                builder.last_progress_time().max(
+                    self.global_sync
+                        .last_progress()
+                        .unwrap_or_else(Timestamp::zero),
```
Could have used `last_progress_time` here too.

rafal-ch replied:
Addressed in 961750e61cca7014966adcf3785324191038f2f7

Commit:
Use the precalculated `last_progress_time`

rafal-ch commented:
bors r+
